File: packages/server/src/repowise/server/schemas/risk_semantics.py
"""Typed wire metadata shared by public risk-report schemas."""

from __future__ import annotations

from typing import Any

from pydantic import BaseModel


class RiskScaleRange(BaseModel):
    minimum: float | None
    maximum: float | None


class RiskCalibration(BaseModel):
    status: str
    source: str | None = None
    calibrated_at: str | None = None
    population: str | None = None
    granularity: str | None = None


class RiskScalarSemantics(BaseModel):
    field: str
    kind: str
    unit: str
    range: RiskScaleRange | None
    measures: str
    deterministic: bool = True
    calibration: RiskCalibration | None = None
    authoritative: bool | None = None
    authoritative_for_change_review: bool | None = None
    runtime_breakage_probability: bool | None = None
    formula: str | None = None
    thresholds: dict[str, float] | None = None
    band_thresholds: dict[str, float] | None = None
    component_fields: dict[str, dict[str, Any]] | None = None


class RiskAuthority(BaseModel):
    authoritative_for: str
    primary_fields: list[str]
    primary_basis: str
    fallback_field: str
    fallback_basis: str
    score_role: str


class RiskCompatibilityField(BaseModel):
    deprecated: bool
    replacement: str
    equivalent_value: bool
    historical_meaning: str

File: packages/types/src/risk-semantics.ts
/** Shared machine-readable vocabulary for public risk and impact scalars. */

export interface RiskScaleRange {
  minimum: number | null;
  maximum: number | null;
}

export interface RiskCalibration {
  status: "benchmarked" | "heuristic_thresholds" | "uncalibrated" | "not_applicable";
  /** Reference tier: present only under include=["scales"]. */
  source?: string | null;
  calibrated_at?: string | null;
  population?: string | null;
  granularity?: string | null;
}

export interface RiskScalarSemantics {
  field: string;
  kind: string;
  unit: string;
  range: RiskScaleRange | null;
  measures: string;
  /** Reference tier: present only under include=["scales"]. */
  deterministic?: boolean;
  calibration?: RiskCalibration | null;
  authoritative?: boolean | null;
  authoritative_for_change_review?: boolean | null;
  runtime_breakage_probability?: boolean | null;
  formula?: string | null;
  thresholds?: Record<string, number> | null;
  band_thresholds?: Record<string, number> | null;
  component_fields?: Record<string, Record<string, unknown>> | null;
}

export interface RiskAuthority {
  authoritative_for: "live_change_review";
  primary_fields: ["risk_percentile", "classification"];
  primary_basis: "benchmarked_population_relative";
  fallback_field: "fallback_band";
  fallback_basis: "absolute_model_score_band";
  score_role: "supporting_diff_shape_signal";
}

export interface RiskCompatibilityField {
  deprecated: true;
  replacement: string;
  equivalent_value: true;
  historical_meaning: string;
}
